### PR TITLE
fix endpoint selection for wildcard to/fromEndpoint in CCNP

### LINKED ISSUE #12844: Empty to/fromEndpoints in CCNP matches world
### Issue
When included in a CNP, an empty `fromEndpoints` is interpreted as "from any Cilium-managed endpoint in the namespace". When included in a **CCNP**, however, an empty `fromEndpoints` is interpreted as "any source" (instead of "all Cilium-managed endpoints").
```
$ cat test.yaml
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "test-all-endpoints"
spec:
  endpointSelector:
    matchLabels:
      name: pod-to-external-fqdn-allow-google-cnp 
  ingress:
  - fromEndpoints:
    - {}
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
$
$ sudo cilium bpf policy get 2506
DIRECTION   LABELS (source:key[=value])                        PORT/PROTO   PROXY PORT   BYTES    PACKETS   
Ingress     reserved:unknown                                   80/TCP       NONE         0        0
[...]
```

### Why?
When converted to `EndpointSelector`, [`to/fromEndpoints` rules are automatically completed with a namespace match if unspecified](https://github.com/cilium/cilium/blob/a24ba16ca353fa63a72978437a702548043dd99d/pkg/k8s/apis/cilium.io/utils/utils.go#L88-L98). However, CCNPs are not namespace scoped so the namespace match isn't added and we end up with a truly empty `EndpointSelector`.

### Proposed Fix
The proposed fix would be to complete empty `to/fromEndpoints` in CCNPs with the following:
```yaml
matchExpressions: {key: io.cilium.kubernetes.namespace, op: exists}
```
This would implement the expected behavior in that empty `to/fromEndpoints` in CCNPs would match all endpoints that have a namespace label, i.e., all Cilium-managed endpoints in the cluster. The fix should be simple (cf. `getEndpointSelector()`) and probably just needs a couple unit tests + documentation update.

### PR DESCRIPTION
**Pull request by fristonio**
See individual commits for more details.

- [x] Fixes #12844 
- [x] Add unit tests.
- [x] Update pre-flight CCNP checks to check for possible required policy upgrade in cluster.

Example of preflight warning.
image

```release-note
Fix endpoint selection for a wildcard to/fromEndpoints in CCNP.
Cilium will only allow access from Cilium-managed endpoints in such cases instead of allowing traffic from any source. Preflight checks, when following the upgrade guide, have been extended to warn users of the new behavior.
```

### CODE REVIEW

**Review by joestringer — COMMENTED** [review]
I was expecting this change to be made in the same (or similar) place where we inject the namespace selector for CNP `EndpointSelector`s, but I don't see that logic here so it's hard to compare against that logic. Were you aware of that other code? Wondering if there's a functional tradeoff from having it here vs. wherever the namespace selector CNP injection logic is.

**Comment by fristonio** [comment]
Hey Joe!
That was our initial approach towards this. But while working on the issue @aanm and @jrajahalme figured that if we introduce the changes where we inject the namespace selector for CNP we would essentially be modifying the semantics of CCNP. So if a user creates a CCNP with the following spec:

```
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "test-all-endpoints"
spec:
  endpointSelector:
    matchLabels:
      app: nginx
  ingress:
  - fromEndpoints:
    - {}
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
```

The policy user gets using `cilium policy get` will look something like this, where we have changed the semantics itself for CCNP provided by the user by adding the matchExpression.

```
[
  {
    "endpointSelector": {
      "matchLabels": {
        "any:app": "nginx"
      }
    },
    "ingress": [
      {
        "fromEndpoints": [
          {
            "matchExpressions": [
              {
                "key": "k8s:io.kubernetes.pod.namespace",
                "operator": "Exists"
              }
            ]
          }
        ],
        "toPorts": [
          {
            "ports": [
              {
                "port": "80",
                "protocol": "TCP"
              }
            ]
          }
        ]
      }
    ],
    "labels": [
      {
        "key": "io.cilium.k8s.policy.derived-from",
        "value": "CiliumClusterwideNetworkPolicy",
        "source": "k8s"
      },
      {
        "key": "io.cilium.k8s.policy.name",
        "value": "test-all-endpoints",
        "source": "k8s"
      },
      {
        "key": "io.cilium.k8s.policy.uid",
        "value": "3f3e0474-3ae7-47e5-8fc2-9eb174ed0594",
        "source": "k8s"
      }
    ]
  }
]
```

This was the reason in the PR, changes were pushed down for a policy update when we resolve the policy for the endpoints rather than injecting it earlier during a parse.

On digging a bit deeper into how this(the current changes) might affect a user we figured that this implementation might also not be correct because we are adding semantics of K8s to a policy that might not even be related to k8s, because at this point we are resolving a general policy rule. So consider a case where user imports the below policy using `cilium policy import $POLICY_FILE`

```
[{
    "endpointSelector": {
      "matchLabels": {
        "any:app": "nginx"
      }
    },
    "ingress": [
      {
        "fromEndpoints": [
          {}
        ],
        "toPorts": [
          {
            "ports": [
              {
                "port": "80",
                "protocol": "TCP"
              }
            ]
          }
        ]
      }
    ],
    "labels": []
}]
```

Even though the policy does not contain any reference to k8s it will still only allow traffic from cilium managed k8s endpoints. So in a non-K8s environment this is essentially blocking all the traffic. 

Now if we were to go back to doing this injection where we do namespace injection for CNP, we will also have the same issue we are trying to solve here for CCNP for all the policies imported via `cilium policy import $POLICY_FILE`. Where we are allowing all traffic for policies with wildcard imported using `cilium policy import`. So what we need is a way to identify all the cilium managed endpoints(irrespective of the environment like K8s) and then add a match for those in the section that is currently in the PR.

We are still discussing on slack what would be the best way to achieve this. I will mark the Pull Request draft for now.

**Comment by joestringer** [comment]
> Now if we were to go back to doing this injection where we do namespace injection for CNP, we will also have the same issue we are trying to solve here for CCNP for all the policies imported via `cilium policy import $POLICY_FILE`.

I think this is an acceptable trade-off. In Kubernetes mode with CNP, we inject the namespace which means that `toEndpoints` / `fromEndpoints` matches only pods. With direct `cilium policy import`, there is no such behaviour so it's "namespace blind" in some sense. Outside of k8s, we're not even guaranteed to have a notion of a 'namespace'. 

If there are users who actively care about this kind of case with non-k8s mode, I would encourage them to look into this topic themselves and come up with a proposal to modify the semantics of `toEndpoints` / `fromEndpoints`. Right now, I don't think there's sufficient user interest in direct policy import for us to address the underlying questions/semantics in that case.

**Comment by fristonio** [comment]
Yeah, that seems like a reasonable trade-off to me as well. So just to be clear we will the do namespace injection in `to/fromEndpoint` selector when we parse the policy from CNP/CCNP right? For direct policy imports, we are currently assuming allow-all as default behavior since we don't have a notion of namespace there?

@jrajahalme @aanm what do you think about this? We will still have the issue we were trying to address earlier, where we change the semantics of CCNP when doing parse.

**Comment by fristonio** [comment]
Pushed the updated fix.

After this patch, a CCNP created with the below configuration
```
$ cat ccnp.yaml
apiVersion: "cilium.io/v2"
kind: CiliumClusterwideNetworkPolicy
metadata:
  name: "test-from-endpoints"
spec:
  endpointSelector:
    matchLabels:
      app: test
  ingress:
  - fromEndpoints:
    - {}
    toPorts:
    - ports:
      - port: "80"
        protocol: TCP
``` 

Will end in the policy repository like this
```
$ cilium policy get
[
  {
    "endpointSelector": {
      "matchLabels": {
        "any:app": "test"
      }
    },
    "ingress": [
      {
        "fromEndpoints": [
          {
            "matchExpressions": [
              {
                "key": "k8s:io.kubernetes.pod.namespace",
                "operator": "Exists"
              }
            ]
          }
        ],
        "toPorts": [
          {
            "ports": [
              {
                "port": "80",
                "protocol": "TCP"
              }
            ]
          }
        ]
      }
    ],
    "labels": [
      {
        "key": "io.cilium.k8s.policy.derived-from",
        "value": "CiliumClusterwideNetworkPolicy",
        "source": "k8s"
      },
      {
        "key": "io.cilium.k8s.policy.name",
        "value": "test-from-endpoints",
        "source": "k8s"
      },
      {
        "key": "io.cilium.k8s.policy.uid",
        "value": "7e204037-e1ec-46aa-9884-41fa1ef339d5",
        "source": "k8s"
      }
    ]
  }
]
```

eventually ending up to allow Ingress from all the endpoints having a namespace label(CIlium managed k8s endpoints), something similar to below output.

```
$ sudo cilium bpf policy get 728
DIRECTION   LABELS (source:key[=value])                                    PORT/PROTO   PROXY PORT   BYTES   PACKETS
Ingress     reserved:host                                                  ANY          NONE         0       0
Ingress     reserved:remote-node                                           ANY          NONE         0       0
Ingress     k8s:io.cilium.k8s.policy.cluster=default                       80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.serviceaccount=cilium-etcd-sa
            k8s:io.cilium/app=etcd-operator
            k8s:io.kubernetes.pod.namespace
Ingress     k8s:app=etcd                                                   80/TCP       NONE         0       0
            k8s:etcd_cluster=cilium-etcd
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=default
            k8s:io.cilium/app=etcd-operator
            k8s:io.kubernetes.pod.namespace
Ingress     k8s:io.cilium.k8s.policy.cluster=default                       80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.serviceaccount=kube-dns
            k8s:io.kubernetes.pod.namespace=kube-system
            k8s:k8s-app=kube-dns
Ingress     k8s:eks.amazonaws.com/component=kube-dns                       80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=kube-dns
            k8s:io.kubernetes.pod.namespace=kube-system
            k8s:k8s-app=kube-dns
Ingress     k8s:io.cilium.k8s.policy.cluster=default                       80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.serviceaccount=coredns
            k8s:io.kubernetes.pod.namespace=kube-system
            k8s:k8s-app=kube-dns
Ingress     k8s:io.cilium.k8s.policy.cluster=default                       80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.serviceaccount=cilium-operator
            k8s:io.cilium/app=operator
            k8s:io.kubernetes.pod.namespace
            k8s:name=cilium-operator
Ingress     k8s:eks.amazonaws.com/component=coredns                        80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=coredns
            k8s:io.kubernetes.pod.namespace=kube-system
            k8s:k8s-app=kube-dns
Ingress     k8s:io.cilium.k8s.policy.cluster=default                       80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.serviceaccount=cilium-etcd-operator
            k8s:io.cilium/app=etcd-operator
            k8s:io.kubernetes.pod.namespace
            k8s:name=cilium-etcd-operator
Ingress     k8s:app=test                                                   80/TCP       NONE         0       0
            k8s:io.cilium.k8s.policy.cluster=default
            k8s:io.cilium.k8s.policy.serviceaccount=default
            k8s:io.kubernetes.pod.namespace=default
Egress      reserved:unknown
```

File: pkg/k8s/apis/cilium.io/utils/utils.go [hunk 1]
```diff
@@ -93,8 +93,17 @@ func getEndpointSelector(namespace string, labelSelector *slim_metav1.LabelSelec
 	// Those pods don't have any labels, so they don't have a namespace label either.
 	// Don't add a namespace label to those endpoint selectors, or we wouldn't be
 	// able to match on those pods.
-	if !matchesInit && !es.HasKey(podPrefixLbl) && !es.HasKey(podAnyPrefixLbl) && namespace != "" {
-		es.AddMatch(podPrefixLbl, namespace)
+	if !matchesInit && !es.HasKey(podPrefixLbl) && !es.HasKey(podAnyPrefixLbl) {
+		if namespace != "" {
```
Comment (joestringer) [inline pkg/k8s/apis/cilium.io/utils/utils.go]: nit: We will usually put the == case first so that when trying to interpret the `else` case we don't have to do the mental swap involved in  `! (namespace != "")`.

**Review by joestringer — CHANGES_REQUESTED** [review]
Looks good to me, just one minor nit.

**Comment by fristonio** [comment]
test-me-please

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 2]
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
+
+	// At this point we have validated the custom resource with the new CRV.
+	// We can try converting it to the new CCNP type.
+	// This should not fail so we are ignoring any errors here.
+	ccnpBytes, err := ccnp.MarshalJSON()
+	if err == nil {
+		resCCNP := cilium_v2.CiliumClusterwideNetworkPolicy{}
+		err = json.Unmarshal(ccnpBytes, &resCCNP)
+		if err != nil {
+			logger.Warnf("Error unmarshalling CCNP bytes: %s", err)
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Please use structured logging here with `WithError()` instead.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 3]
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
+		if resCCNP.Specs != nil {
+			for _, rule := range resCCNP.Specs {
+				if containsEmptyToFromEndpoint(rule) {
+					logger.Warn(warnEmptyToFromEndpointMessage)
+					return nil
+				}
+			}
+		}
+	} else {
+		logger.Warnf("Error marshalling CCNP structure: %s", err)
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Here as well

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 4]
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
 		return errs.ToAggregate()
 	}
+
+	logger := n.logger.WithFields(logrus.Fields{
+		logfields.CiliumClusterwideNetworkPolicyName: ccnp.GetName(),
+	})
+
+	// At this point we have validated the custom resource with the new CRV.
+	// We can try converting it to the new CCNP type.
+	// This should not fail so we are ignoring any errors here.
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: This comment says that we are ignoring errors, but there's a else-clause that logs the error. I wouldn't consider that ignoring :wink:

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 5]
```diff
@@ -105,10 +113,91 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
+
+// containsEmptyToFromEndpoint returns true if a CCNP contains an empty endpoint selector
+// in ingress/egress rules.
+// For more information - https://github.com/cilium/cilium/issues/12844#issuecomment-672074170
+func containsEmptyToFromEndpoint(rule *api.Rule) bool {
+	if len(rule.Ingress) > 0 {
+		for _, r := range rule.Ingress {
+			if len(r.FromEndpoints) > 0 {
+				for _, epSel := range r.FromEndpoints {
+					// Print a warning if the endpoint selector is a wildcard.
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: This comment is doesn't seem necessary anymore.

File: pkg/k8s/apis/cilium.io/v2/validator/validator_test.go [hunk 6]
```diff
@@ -20,11 +20,16 @@ import (
...
 	"testing"
 
+	"github.com/sirupsen/logrus"
 	. "gopkg.in/check.v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 	"sigs.k8s.io/yaml"
 )
 
+var (
+	log = logrus.New()
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator_test.go]: I think we have a convention for instantiating the logger, see other packages.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 7]
```diff
@@ -18,21 +18,27 @@ import (
...
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
 )
 
-// NPValidator is a validator structure used to validate CNP.
+// NPValidator is a validator structure used to validate CNP and CCNP.
 type NPValidator struct {
 	cnpValidator  *validate.SchemaValidator
 	ccnpValidator *validate.SchemaValidator
+
+	logger *logrus.Logger
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Why is this necessary? Can we make it a package-level variable like how we do in other packages throughout the codebase? 

Removing this dependency-injection will also allow us to avoid instantiating the logger in the test code as well.

File: pkg/k8s/apis/cilium.io/v2/validator/validator_test.go (hunk 6 above)
Comment (christarazi) [reply pkg/k8s/apis/cilium.io/v2/validator/validator_test.go]: (Actually, ignore this comment if superseded by the dependency-injection question above.)

**Review by christarazi — CHANGES_REQUESTED** [review]

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 4 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Ah, we are ignoring the error actually as we only log a warning instead of returning the error. But yeah, the comment is not clear. I will update it. :sweat_smile:

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 5 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Oh, I forgot to remove this. :sweat_smile:

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 7 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Yeah, this makes sense actually. We can also attach `validator` subsystem context to the logs. I will update the PR.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 8]
```diff
@@ -105,10 +109,90 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
 	}
+
+	logger := log.WithFields(logrus.Fields{
+		logfields.CiliumClusterwideNetworkPolicyName: ccnp.GetName(),
+	})
+
+	// At this point we have validated the custom resource with the new CRV.
+	// We can try converting it to the new CCNP type.
+	// This should not fail, so we are not returning any errors, just logging
+	// a warning.
```
Comment (christarazi) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go:139]: Thanks for clarifying this. I meant to ask the previous time, but why are we ignoring the errors here? I see that we think that it should not fail, but _if it does_, shouldn't we return an error while logging as well?

**Review by christarazi — COMMENTED** [review]
LGTM; just one question below.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 8 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:139]: I just thought since the validation we are doing here does not lead to an error, we shouldn't throw an error in the process either. But yeah I do think what you are saying makes sense, should I return an error here instead?

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 9]
```diff
@@ -105,10 +109,90 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
 }
 
+const (
+	warnEmptyToFromEndpointMessage = "It seems you have a CiliumClusterwideNetworkPolicy " +
+		"with an empty to/from endpoint selector. The behavior of this selector has been " +
+		"changed. The selector now only allows traffic to/from Cilium managed K8s endpoints, " +
+		"instead of acting as a truly empty endpoint selector allowing all traffic. To " +
+		"ensure that the policy behavior does not affect your workloads, consider adding " +
+		"another policy that allows traffic to/from world and cluster entities. For a more " +
+		"detailed discussion on the topic, see GitHub issue #12844"
```
Comment (pchaigno) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: When will we want to remove this warning? When 1.10 is out? Once 1.9 is the oldest supported version? It would be nice to add a comment such as `We can remove this once we reach 1.x` so we can grep for it later and don't forget to remove it.

**Review by pchaigno — APPROVED** [review]
LGTM. Just one minor nit below.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 9 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go]: This makes sense, I will add a comment regarding this.
One thing I would like to discuss in this context is the backporting scenario for the changes. Do we plan to backport these changes to previous releases? If not we can remove the warnings once the oldest supported version is 1.9

Comment (pchaigno) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Given the upgrade concern, I doubt we can backport this to 1.8/1.7. And unless some users are hitting this, it's probably not worth it anyway.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 8 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:139]: I have the made the changes, let me know if it requires more polishing.

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
test-me-please

**Review by christarazi — APPROVED** [review]

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 10]
```diff
@@ -105,10 +109,89 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
+		return err
+	}
+
+	resCCNP := cilium_v2.CiliumClusterwideNetworkPolicy{}
+	err = json.Unmarshal(ccnpBytes, &resCCNP)
+	if err != nil {
+		return err
+	}
+
+	// Print the warninig only once per CCNP.
```
Comment (joestringer) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go:151]: I'm a little concerned about how frequently this log message could be printed, surely it's not more useful than once per Cilium run? ie wrap with `sync.Once()`?

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 11]
```diff
@@ -105,10 +109,89 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
+		if containsEmptyToFromEndpoint(resCCNP.Spec) {
+			logger.Warning(warnEmptyToFromEndpointMessage)
+			return nil
+		}
+	}
+
+	if resCCNP.Specs != nil {
+		for _, rule := range resCCNP.Specs {
+			if containsEmptyToFromEndpoint(rule) {
+				logger.Warning(warnEmptyToFromEndpointMessage)
```
Comment (joestringer) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: Will this log the warning even if the user intended for it to only select endpoints and not other entities? That's how I at least thought it worked from the beginning :/ Notifying users who are expecting the new behaviour is just extra noise. Would this be more appropriate in a release note / upgrade guide note instead?

**Review by joestringer — COMMENTED** [review]
Looks good overall, only thing I'm wondering about is the log message. The good thing is it tells existing users if their policy is changing. The bad thing is that it'll also give new users a random extra warning potentially once per CCNP parse which might not even be relevant to them.

At most, I think we should print the warning once. I'm on the fence about whether it should be there at all, though I think if it's there and a user hits some issue because of it, at least they would know that this could be why. The opposite side is this could quickly become a false positive warning where an intended policy causes this warning then users get confused when there's not really a problem.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 11 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go]: I think if we are not planning to backport the pull request then it makes sense to remove this extra noise and have this information in the upgrade guide/release note. But if we will be backporting the Pull requests then I think we should not consider that users expect this behavior and have plumbed their policies accordingly. The warning does make sense in such cases. 
 What are your thoughts on this @aanm ?

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 10 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:151]: Yeah, I think it makes sense to print the warning only once to hint the users about the changes. I will update the pull request.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 12]
```diff
@@ -105,10 +109,89 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
 	return nil
 }
+
+// containsEmptyToFromEndpoint returns true if a CCNP contains an empty endpoint selector
+// in ingress/egress rules.
+// For more information - https://github.com/cilium/cilium/issues/12844#issuecomment-672074170
+func containsEmptyToFromEndpoint(rule *api.Rule) bool {
+	if len(rule.Ingress) > 0 {
+		for _, r := range rule.Ingress {
+			if len(r.FromEndpoints) > 0 {
```
Comment (aanm) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go:183]: Why isn't it considered empty if we have other `r.From*` but `r.FromEndpoints == 0`?

**Review by aanm — CHANGES_REQUESTED** [review]

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 12 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:183]: The only place where we do the translation from `LabelSelector` to `EndpointSelector` is for `To/FromEndpoints` and `To/FromRequires`. From the documentation of `FromRequires` 

```
FromRequires is a list of additional constraints which must be met in order for the selected endpoints to be reachable. These additional constraints do no by itself grant access privileges and must always be accompanied with at least one matching FromEndpoints.
```
So we essentially only need to check if `FromEndpoints` contains a wildcard right, or am I missing something here?

Comment (aanm) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:183]: Please add a comment in the code stating :point_up:

**Review by aanm — APPROVED** [review]
There's only the backport concern that we need to address

**Comment by fristonio** [comment]
Rebased the Pull Request.

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 13]
```diff
@@ -105,10 +110,97 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
 	return nil
 }
 
+var (
+	// We can remove the check for this warning once 1.9 is the oldest supported Cilium version.
+	warnWildcardToFromEndpointMessage = "It seems you have a CiliumClusterwideNetworkPolicy " +
+		"with a wildcard to/from endpoint selector. The behavior of this selector has been " +
+		"changed. The selector now only allows traffic to/from Cilium managed K8s endpoints, " +
```
Comment (kkourt) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go:117]:
```suggestion
		"has been changed from 1.9 onwards. The selector now only allows traffic to/from Cilium managed K8s endpoints, " +
```

nit: I think it would be useful to mention from what version behavior changed. Based on the comment above, I'm assuming this changed on 1.9? (feel free to change the phrasing, mine was just a quick suggestion)

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go [hunk 14]
```diff
@@ -105,10 +110,97 @@ func (n *NPValidator) ValidateCNP(cnp *unstructured.Unstructured) error {
...
+		"instead of acting as a truly empty endpoint selector allowing all traffic. To " +
+		"ensure that the policy behavior does not affect your workloads, consider adding " +
+		"another policy that allows traffic to/from world and cluster entities. For a more " +
+		"detailed discussion on the topic, see GitHub issue #12844"
```
Comment (kkourt) [inline pkg/k8s/apis/cilium.io/v2/validator/validator.go]: nit:

```suggestion
		"detailed discussion on the topic, see https://github.com/cilium/cilium/issues/12844"
```

**Review by kkourt — APPROVED** [review]
Just a couple of suggestions for your consideration.
Thanks!

File: pkg/k8s/apis/cilium.io/v2/validator/validator.go (hunk 13 above)
Comment (fristonio) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:117]: We are still discussing the backport scenario for the Pull Request. I will update the comment if required once we decide on that. Thanks for pointing it out though.

Comment (aanm) [reply pkg/k8s/apis/cilium.io/v2/validator/validator.go:117]: We have decided in the Cilium meeting that this PR is going to be backported to 1.7 and 1.8 to avoid having different implementations of the policy in different Cilium versions.

**Comment by fristonio** [comment]
test-me-please
